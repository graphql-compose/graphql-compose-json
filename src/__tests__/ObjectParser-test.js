/* @flow */

import { graphql, TypeComposer } from 'graphql-compose';
import OP from '../ObjectParser';

const { GraphQLFloat, GraphQLBoolean } = graphql;

describe('ObjectParser', () => {
  describe('getFieldConfig()', () => {
    it('number', () => {
      expect(OP.getFieldConfig(6)).toBe('Float');
      expect(OP.getFieldConfig(77.7)).toBe('Float');
    });

    it('string', () => {
      expect(OP.getFieldConfig('test')).toBe('String');
    });

    it('boolean', () => {
      expect(OP.getFieldConfig(true)).toBe('Boolean');
      expect(OP.getFieldConfig(false)).toBe('Boolean');
    });

    it('null', () => {
      expect(OP.getFieldConfig(null)).toBe('JSON');
    });

    describe('array', () => {
      it('of number', () => {
        expect(OP.getFieldConfig([1, 2, 3])).toEqual(['Float']);
      });

      it('of string', () => {
        expect(OP.getFieldConfig(['a', 'b', 'c'])).toEqual(['String']);
      });

      it('of boolean', () => {
        expect(OP.getFieldConfig([false, true])).toEqual(['Boolean']);
      });

      it('of any', () => {
        expect(OP.getFieldConfig([null])).toEqual(['JSON']);
      });
    });

    it('process object', () => {
      const spy = jest.spyOn(OP, 'createTC');
      const valueAsObj = { a: 123 };
      OP.getFieldConfig(valueAsObj, {
        typeName: 'ParentTypeName',
        fieldName: 'subDocument',
      });
      expect(spy).toHaveBeenCalledWith('ParentTypeName_SubDocument', valueAsObj);
    });
  });

  describe('createTC()', () => {
    it('return TypeComposer', () => {
      const tc = OP.createTC('MyType', { a: 1 });
      expect(tc).toBeInstanceOf(TypeComposer);
      expect(tc.getTypeName()).toBe('MyType');
    });

    it('creates fields', () => {
      const tc = OP.createTC('MyType', { a: 1, b: true });
      expect(tc.getFieldNames()).toEqual(['a', 'b']);
      expect(tc.getFieldType('a')).toBe(GraphQLFloat);
      expect(tc.getFieldType('b')).toBe(GraphQLBoolean);
    });

    it('match snapshot', () => {
      const PeopleTC = OP.createTC('PeopleType', {
        name: 'Luke Skywalker',
        height: () => 'Int',
        mass: () => 'Int',
        hair_color: 'blond',
        skin_color: 'fair',
        eye_color: 'blue',
        birth_year: '19BBY',
        gender: 'male',
        homeworld: {
          name: 'Tatooine',
          rotation_period: '23',
          orbital_period: '304',
          terrain: 'desert',
          surface_water: '1',
          population: () => 'Int',
        },
        films: [
          'https://swapi.co/api/films/2/',
          'https://swapi.co/api/films/6/',
          'https://swapi.co/api/films/3/',
        ],
        created: () => 'Date',
        edited: '2014-12-20T21:17:56.891000Z',
      });

      const PeopleGraphQLType = PeopleTC.getType();
      expect(PeopleGraphQLType).toMatchSnapshot();
      expect(PeopleGraphQLType.getFields()).toMatchSnapshot();

      const homeworldSubType = PeopleTC.getFieldType('homeworld');
      expect(homeworldSubType).toMatchSnapshot();
      // $FlowFixMe
      expect(homeworldSubType.getFields()).toMatchSnapshot();
    });
  });
});
