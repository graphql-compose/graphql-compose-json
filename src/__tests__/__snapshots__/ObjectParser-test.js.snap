// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectParser createTC() match snapshot 1`] = `"PeopleType"`;

exports[`ObjectParser createTC() match snapshot 2`] = `
Object {
  "birth_year": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "birth_year",
    "type": "String",
  },
  "created": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "created",
    "type": "Date",
  },
  "edited": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "edited",
    "type": "String",
  },
  "eye_color": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "eye_color",
    "type": "String",
  },
  "films": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "films",
    "type": "[String]",
  },
  "gender": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "gender",
    "type": "String",
  },
  "hair_color": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "hair_color",
    "type": "String",
  },
  "height": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "height",
    "type": "Int",
  },
  "homeworld": Object {
    "args": Array [],
    "description": "",
    "isDeprecated": false,
    "name": "homeworld",
    "type": "PeopleType_Homeworld",
  },
  "mass": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "mass",
    "type": "Int",
  },
  "name": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "name",
    "type": "String",
  },
  "skin_color": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "skin_color",
    "type": "String",
  },
}
`;

exports[`ObjectParser createTC() match snapshot 3`] = `"PeopleType_Homeworld"`;

exports[`ObjectParser createTC() match snapshot 4`] = `
Object {
  "name": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "name",
    "type": "String",
  },
  "orbital_period": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "orbital_period",
    "type": "String",
  },
  "population": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "population",
    "type": "Int",
  },
  "rotation_period": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "rotation_period",
    "type": "String",
  },
  "surface_water": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "surface_water",
    "type": "String",
  },
  "terrain": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "terrain",
    "type": "String",
  },
}
`;
